client/finality-grandpa/rpc/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! RPC API for GRANDPA.
+
+use finality_grandpa::BlockNumberOps;
+use sc_finality_grandpa::{voter, AuthorityId, SharedAuthoritySet, SharedVoterState};
+
+use futures::{FutureExt, TryFutureExt};
+use jsonrpc_derive::rpc;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashSet, fmt::Debug};
+
+const NOT_READY_ERROR_CODE: i64 = 1;
+
+type FutureResult<T> =
+	Box<dyn jsonrpc_core::futures::Future<Item = T, Error = jsonrpc_core::Error> + Send>;
+
+#[derive(derive_more::Display, derive_more::From)]
+pub enum Error {
+	#[display(fmt = "GRANDPA RPC endpoint not ready")]
+	EndpointNotReady,
+}
+
+impl From<Error> for jsonrpc_core::Error {
+	fn from(error: Error) -> Self {
+		jsonrpc_core::Error {
+			message: format!("{}", error).into(),
+			code: jsonrpc_core::ErrorCode::ServerError(NOT_READY_ERROR_CODE),
+			data: None,
+		}
+	}
+}
+
+#[rpc]
+pub trait GrandpaApi {
+	#[rpc(name = "grandpa_roundState")]
+	fn round_state(&self) -> FutureResult<ReportedRoundStates>;
+}
+
+pub struct GrandpaRpcHandler<Hash, Block> {
+	shared_voter_state: SharedVoterState,
+	shared_authority_set: SharedAuthoritySet<Hash, Block>,
+}
+
+impl<Hash, Block> GrandpaRpcHandler<Hash, Block> {
+	pub fn new(
+		shared_voter_state: SharedVoterState,
+		shared_authority_set: SharedAuthoritySet<Hash, Block>,
+	) -> Self {
+		Self {
+			shared_voter_state,
+			shared_authority_set,
+		}
+	}
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Prevotes {
```

Comment:
Why are `Prevotes`, `Precommits`, `RoundState` and all their fields public?

Reply:
Yeah good point. I went through and removed `pub` in a number of places!